tests/unit/s2n_random_test.c

```diff
@@ -791,27 +793,41 @@ static int s2n_random_rand_bytes_after_cleanup_cb(struct random_test_case *test_
     return S2N_SUCCESS;
 }
 
+/* OpenSSL's RAND_METHOD->bytes function pointer type.
+ *
+ * AWS-LC has a different function signature which results in a compiler error.
+ * We cast to this common function pointer type **only when comparing** in
+ * tests.
+ *
+ * It is undefined behavior to cast a function pointer to a function pointer of
+ * a different type and then call that function.
+ *
+ * - AWS-LC: https://github.com/aws/aws-lc/blob/main/include/openssl/rand.h#L124
+ * - OpenSSL: https://github.com/openssl/openssl/blob/master/include/openssl/rand.h#L42
+ */
+typedef int (*s2n_rand_meth_fn_type)(unsigned char *buf, int num);
+
 static int s2n_random_rand_bytes_before_init(struct random_test_case *test_case)
 {
-#if S2N_LIBCRYPTO_SUPPORTS_CUSTOM_RAND
-    /* Calling RAND_bytes will set a global random method */
-    unsigned char rndbytes[16] = { 0 };
-    EXPECT_EQUAL(RAND_bytes(rndbytes, sizeof(rndbytes)), 1);
-    const RAND_METHOD *rand_method = RAND_get_rand_method();
-    EXPECT_NOT_NULL(rand_method);
-    EXPECT_NOT_EQUAL(rand_method->bytes, s2n_openssl_compat_rand);
-
-    EXPECT_SUCCESS(s2n_init());
+    /* s2n_query_libcrypto_is_fips() is used since we are testing `s2n_init()` */
+    if (s2n_supports_custom_rand() && !s2n_query_libcrypto_is_fips()) {
+        /* Calling RAND_bytes will set a global random method */
+        unsigned char rndbytes[16] = { 0 };
+        EXPECT_EQUAL(RAND_bytes(rndbytes, sizeof(rndbytes)), 1);
+        const RAND_METHOD *rand_method = RAND_get_rand_method();
+        EXPECT_NOT_NULL(rand_method);
+        EXPECT_NOT_EQUAL((s2n_rand_meth_fn_type) rand_method->bytes, (s2n_rand_meth_fn_type) s2n_openssl_compat_rand);
```

Comment:
Defining a type for this is probably overkill. I'd forgotten that you can't use `void*` for functions and there's no function version of `uintptr_t` -_- Still, you should be able to just do:
```suggestion
        EXPECT_NOT_NULL(rand_method);
        EXPECT_NOT_EQUAL((void (*)(void)) rand_method->bytes, (void (*)(void)) s2n_openssl_compat_rand);
```
It just looks sillier than `void*` or `uintptr_t`.

Reply:
Originally I used `void*` but that resulted in valgrind warnings about casting to void*. We could wrap the cast in lint ignores but that would be uglier imo. Not sure if there is a better option.